### ISSUE TITLE
[ethers] Generate constructor passing ABI to ethers

### ISSUE THREAD

**Issue by Marik-D** [ethereum-ts/TypeChain#137]
Hi, 

right now ethers target only generates type definitions, leaving the user with need to manually instantiate the ethers `Contract` class, passing the ABI to it, and then cast the result to a correct type.
So, basically,  right now something like this code is required:
```typescript
import { Contract } from 'ethers';
import { Token } from 'token-contract/types/Token';
import { TokenABI } from 'token-contract/abi/Token.json';
...
const token: Token = new Contract(address, TokenABI, provider) as Token;
```
It would be nice, if `ethers` target generated typescript files having constructor implemented which passed contract ABI to base class (similar how legacy target does it now).
So, new generated code would look something like:
```typescript
const ABI = [
  ...
  <abi>
  ...
];

export class Agreement extends Contract {
  constructor(nameOrAddress: string | Address, providerOrSigner: Provider | Signer) {
    super(nameOrAddress, ABI, providerOrSigner);
  }

  functions: {
     ...
   }
}
```
This would ease the use of generated classes, making only this implementation needed to be exported from package (instead of type definitions and the ABI).

P.S. 
We would love to implement that kind generation in [Waffle](https://getwaffle.io) to work out-of-the-box.

**Comment by krzkaczor** [comment]
Not generating anything other than pure types is a conscious design choice for TypeChain. It greatly minimizes API surface since we just recreate whatever is done by external library authors. 

That being said I think this could be solved by user-defined targets which at the time are not supported :( I am gonna leave this open for now.

**Comment by quezak** [comment]
I took a quick look at ethers v5 beta and they indeed added TS classes generation there, with the ABI and bytecode built-in, but typings for other parts of the contract are much simpler than TypeChain's for now. Also, v5 still has no docs, so it's not ready to be used yet. I may take another look, and even port some of our typings there, but for now I created PR #160 to introduce deployments to TypeChain too.
